README.md
Insaned
=======

Insaned is a simple linux daemon for polling button presses on SANE scanners.

[![Build Status](https://travis-ci.org/abusenius/insaned.svg?branch=master)](https://travis-ci.org/abusenius/insaned) [![Coverity Scan Build Status](https://scan.coverity.com/projects/11384/badge.svg)](https://scan.coverity.com/projects/abusenius-insaned)

Description
-----------

Insaned periodically polls your scanner using the SANE library and runs the corresponding event handler script when a button is pressed. Because of this, button presses are only detected every N milliseconds, so you will have to press and hold the button for at most N milliseconds until an event is fired. Insaned does only fire the same event once in 2500 ms to prevent unwanted repetitions.

It should work with all backends that expose buttons as "Sensors". The daemon reads the value of all sensors every N milliseconds (default: 500) and starts an event handler script named by the sensor name. Polling does not result in a noticeable CPU load, but produces some I/O load. Therefore, it might not be a good idea to run this daemon on a laptop, since it will probably prevent USB bus from entering a low power mode or even keep the laptop awake (not tested yet).

Currently, insaned was tested on:
* Gentoo Linux with sane-backends-1.0.24 and a Canon LiDE 210 flatbed scanner (genesys backend, USB ID 04a9:190a)
* Ubuntu 16.04.1 LTS with Canon LiDE 210
* Raspberry Pi with HP Scanjet 2400c (thanks to [GaryA](https://github.com/GaryA))
* Debian 8 running on a [CHIP](https://getchip.com/pages/chip) with Canon Lide 210 (thanks to [Coaxial](https://github.com/Coaxial))
* FreeBSD 11.1 with Fujitsu ScanSnap S1500 (thanks to [displaced](https://github.com/displaced))


Why
---

I've tried [scanbuttond](http://scanbuttond.sourceforge.net/) and [scanbd](http://scanbd.sourceforge.net/).

* scanbuttond could not detect buttons on my scanner
* scanbd is waaay too complicated, it requires to configure SANE for network scanning. I spend 2 evenings trying to make it work and failed.
* I've heard rumors about sanebuttonsd, but couldn't find any documentation about it


Supported devices
-----------------

You can check if SANE exposes sensors of your scanner by running:

    scanimage -A

You should be able to see a list of detected sensors and their current state. If you press and hold a button while running scanimage -A, you should be able to see a different sensor value.

For example, this is the output for my scanner, when I press the "file" button (note the [yes]):

    All options specific to device `genesys:libusb:001:003':
      Scan Mode:
    ...
    <snip>
    ...
      Sensors:
        --scan[=(yes|no)] [no] [hardware]
            Scan button
        --file[=(yes|no)] [yes] [hardware]
            File button
        --email[=(yes|no)] [no] [hardware]
            Email button
        --copy[=(yes|no)] [no] [hardware]
            Copy button
        --extra[=(yes|no)] [no] [hardware]
            Extra button
      Buttons:
        --clear-calibration
            Clear calibration cache

In this case, insaned detects all of my 5 buttons and emits events named "scan", "file", "email", "copy" and "extra".


Event Handler Scripts
---------------------

Event handler scripts are simple shell scripts. Insaned searches for them in /etc/insaned/events/ directory (configurable). The daemon passes current SANE device name as the first and only argument to the script, in case you need to distinguish between several scanners.

All event handler scripts have to exist and have to have the executable flag set, otherwise insaned will print warnings. Create an empty executable file to silence the warning, e.g. like this:

    touch /etc/insaned/events/scan
    chmod +x /etc/insaned/events/scan


Dependencies
------------

* SANE library version 1.0.23 or later (tested with: 1.0.23 - 1.0.27)
* Recent GCC with support for C++11 standard (tested with: 4.8.2, 5.4.0, 6.4.0)


Installation
------------

To compile insaned, run:

    make

in the project directory. The daemon will be created in the project directory and called "insaned". You can run it in foreground for testing purposes as follows:

    ./insaned --dont-fork --events-dir=$PWD/events --log-file=$PWD/log.log -vv

See also

    ./insaned --help --list-sensors

for more details.

*Tips and tricks*

If you happen to have a system where SANE headers (sane/sane.h) and libraries (libsane.so) are installed in an unusual location and simple `make` fails to compile insaned, try to provide paths to headers and libraries as follows:

    CXXFLAGS=-I/path/to/your/usr/include LDFLAGS=-L/path/to/your/usr/lib make

*Gentoo Linux*

1. Add
   [media-gfx/insaned/insaned-0.0.3.ebuild](https://raw.githubusercontent.com/abusenius/insaned/master/gentoo/media-gfx/insaned/insaned-0.0.3.ebuild) to media-gfx/insaned in your local overlay.
2. cd $YOUR_OVERLAY/media-gfx/insaned
3. ebuild insaned-0.0.3.ebuild manifest
4. emerge insaned

*Ubuntu Linux (64 bit)*

1. Download [insaned_0.0.3-0ubuntu1_amd64.deb](https://github.com/abusenius/insaned/releases/download/v0.0.3/insaned_0.0.3-0ubuntu1_amd64.deb)
2. sudo dpkg -i insaned_0.0.3-0ubuntu1_amd64.deb

*FreeBSD*

1. Use `gmake` instead of `make` to compile insaned.
2. If `gmake` also doesn't work, see *Tips and tricks* above
3. In order to automatically start insaned when the scanner is connected or turned on and stop it when the scanner is disconnected/turned off use provided [devd/insaned.conf](https://raw.githubusercontent.com/abusenius/insaned/master/freebsd/devd/insaned.conf.example) example


Side note
---------

I am not responsible for any physical or mental damage insaned might cause to you, your hardware or your pet. Use it on your own risk.

That said, any help is welcomed. Feel free to contact me if you have any comments/suggestions/patches. Especially interesting are reports if it works with another scanners. Please make sure to include the output of "insaned -L" to your report, and, if you have any problems, also the output of "scanimage -A".


TODO
----

The following features are planned:

* Package for raspbian
* Suspend polling while one of the configured processes are running
* Suspend polling while another process is using the SANE library
* CMake build
* Packages for other linux distributions
* More useful handler scripts
* Handler script stub for starting a GUI frontend
* Run daemon as user daemon when logging into a KDE/Gnome/whatever
* Test with more hardware



debian/changelog
insaned (0.0.3-0ubuntu1) xenial; urgency=low

  * Autostart systemd service after installation
  * Updated event scripts, added multipage scanning example

 -- Alex Busenius <abusenius@users.noreply.github.com>  Sat, 28 Oct 2017 23:18:50 +0200

insaned (0.0.2-0ubuntu1) xenial; urgency=medium

  * Initial release

 -- Alex Busenius <abusenius@users.noreply.github.com>  Thu, 18 Aug 2016 21:08:42 +0200


src/insaned.cpp
/* insaned -- simple daemon polling button presses using the SANE library
   Heavily based on scanimage utility from SANE distribution
   Copyright (C) 2013 - 2016 Alex Busenius

   scanimage -- command line scanning utility
   Uses the SANE library.
   Copyright (C) 1996, 1997, 1998 Andreas Beck and David Mosberger

   Copyright (C) 1999 - 2009 by the SANE Project

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License as
   published by the Free Software Foundation; either version 2 of the
   License, or (at your option) any later version.

   This program is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.  */

#include "config.h"

#include <iostream>
#include <string>
#include <getopt.h>
#include <syslog.h>
#include <unistd.h>
#include <cstdio>
#include <cstdlib>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <cstring>
#include <cerrno>
#include <cstring>
#include <cstdio>

#include "InsaneDaemon.h"
#include "InsaneException.h"


std::string basename(const std::string & path)
{
    if (path.empty()) {
        return "";
    }
    auto pos = path.rfind("/");
    if (pos == std::string::npos) {
        return "";
    }
    return path.substr(0, pos);
}


bool createIfNeeded(const std::string & path, bool isFile)
{
    struct stat sb;
    const mode_t mode = S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH | (isFile ? 0 : S_IXUSR | S_IXGRP | S_IXOTH);

    if (isFile) {
        int fd = open(path.c_str(), O_CREAT | O_WRONLY, mode);
        if (fd == -1) {
            if (errno != EISDIR) {
                syslog(LOG_ERR | LOG_USER, "Creat failed: %s", strerror(errno));
                return false;
            }
        } else {
            close(fd);
        }
    } else {
        if (mkdir(path.c_str(), mode) == -1) {
            if (errno != EEXIST) {
                syslog(LOG_ERR | LOG_USER, "Mkdir failed: %s", strerror(errno));
                return false;
            }
        }
    }

    if (chmod(path.c_str(), mode) == -1) {
        if (errno != ENOENT) {
            syslog(LOG_ERR | LOG_USER, "Chmod failed: %s", strerror(errno));
            return false;
        }
    }

    if (stat(path.c_str(), &sb) == -1) {
        syslog(LOG_ERR | LOG_USER, "Stat failed: %s", strerror(errno));
        return false;
    }

    if (!isFile && (sb.st_mode & S_IFMT) != S_IFDIR) {
        syslog(LOG_ERR | LOG_USER, "%s is not a directory", path.c_str());
        return false;
    }
    if (isFile && (sb.st_mode & S_IFMT) != S_IFREG) {
        syslog(LOG_ERR | LOG_USER, "%s is not a regular file", path.c_str());
        return false;
    }

    return true;
}


int main(int argc, char ** argv)
{
    // defaults
    const std::string LOGFILE       = "/var/log/" + InsaneDaemon::NAME + ".log";
    const std::string EVENTS_DIR    = "/etc/" + InsaneDaemon::NAME + "/events";
    const int SLEEP_MS              = 500;
    const int SLEEP_MIN             = 50;
    const int SLEEP_MAX             = 5000;
    const int VERBOSITY             = 0;
    const bool DO_FORK              = true;
    const bool SUSPEND_AFTER_EVENT  = false;

    // command line options
    const char * BASE_OPTSTRING = "d:hvVf:e:s:nLwp:";
    option basic_options[] = {
        {"device-name", required_argument, nullptr, 'd'},
        {"help", no_argument, nullptr, 'h'},
        {"verbose", no_argument, nullptr, 'v'},
        {"version", no_argument, nullptr, 'V'},
        {"log-file", required_argument, nullptr, 'f'},
        {"events-dir", required_argument, nullptr, 'e'},
        {"sleep-ms", required_argument, nullptr, 's'},
        {"dont-fork", no_argument, nullptr, 'n'},
        {"list-sensors", no_argument, nullptr, 'L'},
        {"suspend-after-event", no_argument, nullptr, 'w'},
        {"pid-file", required_argument, nullptr, 'p'},
        {0, 0, nullptr, 0}
    };

    std::string prog_name;
    prog_name = std::string(argv[0]);
    size_t n = prog_name.rfind('/');
    if (n != std::string::npos) {
        prog_name = prog_name.substr(n + 1);
    }

    bool help = false;
    bool list = false;
    bool suspend = SUSPEND_AFTER_EVENT;
    int verbose = VERBOSITY;
    bool do_fork = DO_FORK;
    int sleep_ms = SLEEP_MS;
    std::string devname = "";
    std::string pidfile = "";
    std::string logfile = LOGFILE;
    std::string events_dir = EVENTS_DIR;

    // get dameon instance
    InsaneDaemon & daemon = InsaneDaemon::instance();

    /* make a pass through the options with error printing */
    int ch = 0;
    int index = 0;
    opterr = 1;
    while ((ch = getopt_long(argc, argv, BASE_OPTSTRING, basic_options, &index)) != EOF) {
        switch (ch) {
        case ':':
        case '?':
            return 1; // error is printed by getopt_long
            break;
        case 'd':
            devname = optarg;
            break;
        case 'h':
            help = true;
            break;
        case 'L':
            list = true;
            break;
        case 'v':
            verbose++;
            break;
        case 'V':
            std::cout << InsaneDaemon::NAME << " " << VERSION
                      << ". SANE backend version " << daemon.get_sane_version() << std::endl;
            return 0;
        case 'f':
            logfile = optarg;
            break;
        case 'e':
            events_dir = optarg;
            break;
        case 'p':
            pidfile = optarg;
            break;
        case 's':
            try {
                sleep_ms = std::stoi(std::string(optarg));
                if (sleep_ms < SLEEP_MIN || SLEEP_MAX < sleep_ms) {
                    throw std::out_of_range("The value must be in range " + std::to_string(SLEEP_MIN) + ".." + std::to_string(SLEEP_MAX));
                }
            } catch (std::exception & e) {
                std::cerr << "Invalid value of --sleep-ms (" << optarg << "): " << e.what() << std::endl;
                return 1;
            }
            break;
        case 'n':
            do_fork = false;
            break;
        case 'w':
            suspend = true;
            break;
        default:
            std::cerr << "Unknown option: " << static_cast<char>(ch) << std::endl;
            return 1;
            break;
        }
    }

    try {
        daemon.init(devname, events_dir, sleep_ms, verbose, do_fork && !(help || list), suspend);

        /* print help and device list */
        if (help) {
            std::cout << "Usage: " << prog_name << " [OPTIONS]...\n"
                << "\n"
                << "Start polling sensors of a scanner device and run the corresponding callback\n"
                << "script when a button is pressed.\n"
                << "\n"
                << "Parameters are separated by a blank from single-character options (e.g.\n"
                << "-d epson) and by a \"=\" from multi-character options (e.g. --device-name=epson).\n"
                << " -d, --device-name=DEVICE   use the given scanner device instead of the first\n"
                << "                            available device\n"
                << " -f, --log-file=FILE        use the given log file instead of default\n"
                << "                            (" << LOGFILE << ")\n"
                << " -e, --events-dir=DIR       execute event scripts from the given directory\n"
                << "                            instead of the default (" << EVENTS_DIR << ")\n"
                << " -s, --sleep-ms=NUMBER      sleep for the given amount of ms between polling the\n"
                << "                            sensors instead of default (" << SLEEP_MS << " ms), must be in\n"
                << "                            range " << SLEEP_MIN << ".." << SLEEP_MAX << "\n"
                << " -n, --dont-fork            do not fork into background\n"
                << " -L, --list-sensors         list sensors that will be monitored along with their\n"
                << "                            current state and exit. See also --device-name\n"
                << " -w, --suspend-after-event  suspend sensor polling for 15 seconds after an event\n"
                << "                            handler script was triggered. Use this if insaned\n"
                << "                            tends to interfere with your handlers.\n"
                << " -p, --pid-file=FILE        if this option is present, the daemon will create\n"
                << "                            this file and write its PID into it after fork\n"
                << " -v, --verbose              give even more status messages\n"
                << " -h, --help                 display this help message and exit\n"
                << " -V, --version              print version information and exit" << std::endl;

            std::cout << "\nList of available devices:\n";
            for (auto & device : daemon.get_devices()) {
                std::cout << "    " << device << std::endl;
            }
            if (!list) {
                return 0;
            }
            std::cout << std::endl;
        }

        if (list) {
            auto sensors = daemon.get_sensors(); // updates current device if it was not set
            std::cout << "List of sensors for device '" << daemon.current_device() << "':" << std::endl;
            for (auto & pair : sensors) {
                std::cout << "    " << pair.first << "\t" << (pair.second ? "[yes]" : "[no]") << std::endl;
            }
            return 0;
        }
    } catch (InsaneException & e) {
        std::cerr << InsaneDaemon::NAME << ": " << e.what() << std::endl;
        return 1;
    } catch (std::exception & e) {
        std::cerr << InsaneDaemon::NAME << ": " << e.what() << std::endl;
        return 1;
    }

    try {
        // create and chmod logfile and pid directory
        if (!createIfNeeded(logfile, true)) {
            return 1;
        }
        if (!pidfile.empty() && !createIfNeeded(basename(pidfile), false)) {
            return 1;
        }

        // TODO chown logfile and pid directory
        // TODO su to another UID
        // TODO drop permissions
        if (do_fork) {
            // fork
            if (pid_t pid = fork()) {
                if (pid > 0) {
                    // parent process
                    return 0;
                } else {
                    syslog(LOG_ERR | LOG_USER, "First fork failed: %s", strerror(errno));
                    return 1;
                }
            }

            // detach
            setsid();

            // some settings
            if (chdir("/") < 0) {
                syslog(LOG_ERR | LOG_USER, "Unable to chdir to /: %s", strerror(errno));
                return 1;
            }
            umask(0);

            // second fork ensures the process cannot acquire a controlling terminal.
            if (pid_t pid = fork()) {
                if (pid > 0) {
                    // parent
                    return 0;
                } else {
                    syslog(LOG_ERR | LOG_USER, "Second fork failed: %s", strerror(errno));
                    return 1;
                }
            }

            // close streams
            close(0);
            close(1);
            close(2);

            // disable standard input.
            if (open("/dev/null", O_RDONLY) < 0) {
                syslog(LOG_ERR | LOG_USER, "Unable to open /dev/null: %s", strerror(errno));
                return 1;
            }

            // send standard output to a log file
            const int flags = O_WRONLY | O_CREAT | O_APPEND;
            const mode_t mode = S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH;
            if (open(logfile.c_str(), flags, mode) < 0) {
                syslog(LOG_ERR | LOG_USER, "Unable to open log file %s: %s", logfile.c_str(), strerror(errno));
                return 1;
            }

            // send standard error to the same log file
            if (dup(1) < 0) {
                syslog(LOG_ERR | LOG_USER, "Unable to dup output descriptor: %s", strerror(errno));
                return 1;
            }

            if (!pidfile.empty()) {
                int fid = open(pidfile.c_str(), O_WRONLY | O_CREAT | O_TRUNC, mode);
                if (fid < 0) {
                    syslog(LOG_ERR | LOG_USER, "Unable to open pid file %s: %s", pidfile.c_str(), strerror(errno));
                    return 1;
                }
                FILE * f = fdopen(fid, "w");
                if (f) {
                    fprintf(f, "%d\n", getpid());
                    fclose(f);
                }
                close(fid);
                f = nullptr;
            }

            syslog(LOG_INFO | LOG_USER, "Daemon started");
        }

        // run daemon
        daemon.run();

        if (do_fork) {
            syslog(LOG_INFO | LOG_USER, "Daemon stopped");
        }

        return 0;
    } catch (InsaneException & e) {
        syslog(LOG_ERR | LOG_USER, "Exception: %s", e.what());
        std::cerr << "\n" << InsaneDaemon::NAME << ": Exception: " << e.what() << std::endl;
    } catch (std::exception & e) {
        syslog(LOG_ERR | LOG_USER, "Exception: %s", e.what());
        std::cerr << "\n" << InsaneDaemon::NAME << ": Exception: " << e.what() << std::endl;
    }

    return 1;
}

